ykrt/src/location.rs

```diff
@@ -334,9 +321,25 @@ impl Drop for Location {
         let ls = self.load(Ordering::Relaxed);
         if !ls.is_counting() {
             debug_assert!(!ls.is_locked());
-            let hr = unsafe { ls.hot_location() };
-            unsafe {
-                Box::from_raw(hr);
+            self.lock().unwrap();
+            let hl = unsafe { ls.hot_location() };
+            if let HotLocation::Compiled(_) = hl {
+                // FIXME: we can't drop this memory as another thread may still be executing the
+                // trace that's pointed to. There should be a ref count that we decrement and free
+                // memory if it reaches zero.
+                self.unlock();
+            } else if let HotLocation::Compiling = hl {
+                // Another thread has a pointer to the underlying `HotLocation`, so we have to
+                // signal to that thread that they need to free the `HotLocation`.
+                *hl = HotLocation::Dropped;
+                self.unlock();
+            } else if let HotLocation::DontTrace | HotLocation::Tracing(_) = hl {
+                self.unlock();
```

Comment:
Are the repeated unlocks necessary? I assume you are trying to make the critical section small?

Reply:
> I assume you are trying to make the critical section small?

Exactly! Since we know that `Location`s will be highly contended, reducing the critical sections is very important.